### DISCUSSION TITLE
[Question] DifferentialIKController cannot reach the goal for the Booster T1 robot.

### DISCUSSION THREAD

**Discussion by InnaSviri** [isaac-sim/IsaacLab#5073] [Q&A]
### Question

I am working on generating a synthetic dataset to control the T1 booster's hands. I would like to use GPU-oriented IK to integrate this data generation with RL training in joint space. For these purposes, I’m trying to use DifferentialIKController. 

Similarly to DifferentialInverseKinematicsAction, I obtain the desired joints from ik_controller.compute at the simulation frequency and set the command to ik_controller at the policy frequency. I have sampled the end-effector gains, interpolating between the start and end points within the following ranges:
                "x_start": (0.3, 0.35),
                "y_start": (-0.3, 0.1),
                "z_start": (-0.1, 0.1),
                "roll_start": (0., 0.),
                "pitch_start": (0., 0.),
                "yaw_start": (0., 0.),
                "x_end": (0.3, 0.35),
                "y_end": (-0.3, 0.1),
                "z_end": (-0.1, 0.1),
                "roll_end": (0., 0.),
                "pitch_end": (0., 0.),
                "yaw_end": (0., 0.),

However, the decisions obtained from the ik_controller output are unreliable (see [video](https://1drv.ms/v/c/df93b9afd6f38bf0/IQCtVAM5LewKR6Vec9iQ35HHAedIxT1Y_GG9Uc0OR3BmWks?e=oV4fdo)). The pinocchio IK handles this ranges well.

The situation is slightly better with more limited sampling ranges, but it is still not good enough for training a policy on it ([video](https://1drv.ms/v/c/df93b9afd6f38bf0/IQCroEZpIUDbRLeMzbuLDA6wAUUksJRfVsXWq2sPB8Qvgbw?e=o8fZDq)).
                "x_start": (0.2, 0.25),
                "y_start": (-0.3, -0.1),
                "z_start": (-0.1, 0.1),
                "roll_start": (0., 0.),
                "pitch_start": (0., 0.),
                "yaw_start": (0., 0.),
                "x_end": (0.2, 0.25),
                "y_end": (-0.3, -0.1),
                "z_end": (-0.1, 0.1),
                "roll_end": (0., 0.),
                "pitch_end": (0., 0.),
                "yaw_end": (0., 0.),
                
Could someone give me an idea of how to solve this problem?

Thank you in advance!

### Build Info

Describe the versions that you are currently using:

Isaac Lab Version: v2.2.1
Isaac Sim Version: 4.5

**Comment by RandomOakForest** [comment]
Thank you for posting this. Would it be possible for you to upgrade to Isaac Sim 5.1 and Isaac Lab v2.3.2?

**Comment by InnaSviri** [comment]
Thank you for your response! I have upgraded to Isaac Sim 5.1 and Isaac Lab v2.3.2, but the issue still persists - [video](https://1drv.ms/v/c/df93b9afd6f38bf0/IQAHOMzzrSyDQL8eo2K7mqPcASiWxPxk45xDWLg1C9QpNJ8?e=QkBh0J)

Perhaps you could suggest another way to generate a synthetic dataset for controlling the T1 booster's hands during joint space training?

**Comment by RandomOakForest** [comment]
Thank you for following up and upgrading your toolset. The behavior you’re seeing is almost always caused by a **mismatch between how DifferentialIKController is configured/used and how you’re generating the task‑space commands and Jacobians**, not by the workspace ranges themselves.[^1][^2][^3]

Below are the concrete things I would check and adjust. I'll move this post to our _Discussions_ section for follow-up.


## 1. Command type and relative mode

DifferentialIKController can interpret the `command` either as an absolute pose or as a delta pose, and that must match how you build your trajectory.[^2][^1]

- If `command_type="pose"` and `use_relative_mode=False`, then:
    - `command[..., :3]` (if you use position‑only) should be the **absolute desired EE position** in the same frame as `ee_pos` passed into `compute`.
    - In that case, you should **not** be interpolating in “delta space” and feeding deltas; you should interpolate absolute poses and pass them directly.
- If `command_type="pose_rel"` and/or `use_relative_mode=True`, then:
    - The `command` must be a **delta pose**: $\Delta x, \Delta y, \Delta z$ (and rotation delta) over a controller time step.[^3]
    - These deltas should scale with the controller timestep (e.g., per‑step increment of your Cartesian trajectory), not the full start→end displacement.

If you are:

- computing a full start→end displacement and directly giving it as the command every step, or
- providing absolute positions but configuring the controller for relative mode,

you will see exactly the kind of overshoot, oscillation or inconsistent solutions that make the output unusable for RL.

**Actionable check**

- Print/inspect:
    - `diff_ik_cfg.command_type`
    - `diff_ik_cfg.use_relative_mode`
- Make them consistent with how you build `command`:
    - For trajectory in absolute Cartesian: `command_type="pose"`, `use_relative_mode=False`, `command = x_interp(t)` (no deltas).
    - For per‑step deltas: `command_type="pose_rel"`, `use_relative_mode=True`, `command = x_interp(t+dt) - x_interp(t)`.



## 2. Jacobian and frame consistency

DifferentialIKController assumes that:

- `ee_pos` and `ee_quat` are the **current EE pose in some task frame** (often world or base).[^1]
- `jacobian` is the **geometric Jacobian mapping joint velocities to spatial twist** in the **same frame and convention** as the pose.
- The order of the joints in `jacobian[..., :, :]` matches `joint_pos` exactly.

If any of these are off, the solver can behave badly even though Pinocchio IK works (Pinocchio may use a different reference frame and its own Jacobian building).

Key pitfalls to check:

- Are you using the EE Jacobian in the same frame as the pose you pass into `compute` (e.g., world frame vs base frame)?
- Does your Jacobian order match the joint DOF order used in the articulation?
- Are you accidentally passing a body Jacobian that’s not aligned with the same origin/orientation as `ee_pos` / `ee_quat`?

**Minimal sanity test**

At a fixed configuration:

1. Set `command` equal to the **current pose** (absolute mode) or zero delta (relative mode).
2. Run `compute` for several steps while holding the command constant.
3. The joint targets should converge to a fixed configuration with almost no drift.
    - If they drift or vibrate, you likely have a frame or scaling mismatch.



## 3. IK method and parameters (damping, gains)

DifferentialIKController supports several Jacobian inversion methods and parameters that affect stability and workspace behavior.[^4][^5][^3][^1]

- For manipulators working over a wide workspace, I strongly recommend:
    - `ik_method="dls"` with a moderate `lambda_val` (e.g. `1e-2` to `5e-2`).
- If you use `"pinv"` or `"svd"` without damping, you can see wild swings near singularities or ill‑conditioned regions.
- There is also a **scaling factor** on the joint updates (`k_val` for some methods) which effectively acts like a Cartesian proportional gain; too large and you get overshoot/excitation.

**Actionable configuration**

For example:

```python
from isaaclab.controllers import DifferentialIKController, DifferentialIKControllerCfg

diff_ik_cfg = DifferentialIKControllerCfg(
    command_type="pose",           # or "pose_rel"
    use_relative_mode=False,       # match how you form commands
    ik_method="dls",
    ik_params={"lambda_val": 0.02} # try 0.01–0.05
)
diff_ik = DifferentialIKController(
    cfg=diff_ik_cfg,
    num_envs=num_envs,
    device=sim.device
)
```

If you already use `"dls"`, try increasing `lambda_val` and slightly reducing the magnitude of your per‑step delta pose.



## 4. Multi‑rate control: sim vs policy frequency

You mentioned:

> compute at simulation frequency and set the command at policy frequency.

DifferentialIKController is a **differential** method; it expects that the relationship between command increments and integration steps is consistent.[^5][^3]

Common failure mode:

- You compute joint targets each sim step assuming a small $\Delta x$, but only update the high‑level command at a slower policy frequency while using large deltas that effectively span multiple sim steps.

Best practice:

- Let the IK controller run at **the same frequency as the pose trajectory**.
- If policy runs at a lower rate, do one of:
    - Make the policy output a **desired pose** (absolute), and have a local interpolator (in task space) generate the per‑step `command` at the sim frequency before passing to DifferentialIKController.
    - Or, if output is a delta, ensure it is **normalized per sim step** (divide by number of sim steps per policy step) so that each sim step only sees a fraction of the movement.

Also confirm that you’re not repeatedly applying the same large delta at each sim step between policy updates.



## 5. Workspace and joint limits vs Pinocchio IK

Pinocchio IK (or PinkIKController) usually does:

- optimization with joint limit and sometimes posture regularization explicitly considered.[^5][^1]

DifferentialIKController by default:

- integrates joint updates without explicit optimization, relying on damping and possibly clipping to joint limits.

If your T1 booster hand is near joint limits or singular postures in the sampled regions:

- Pinocchio can “bend around” those constraints.
- Differential IK can hit limits, saturate, and produce poor trajectories.

Mitigations:

- Ensure **joint limit clipping** is enabled in your higher‑level control (clip `joint_pos_target` before sending).
- Slightly shrink the task‑space ranges away from singular zones or extreme extension.
- Increase damping (`lambda_val`) and potentially reduce per‑step Cartesian velocity.



## 6. Quick diagnostic experiment

To isolate the issue away from sampling ranges:

1. Pick a single start pose that Pinocchio IK can reach comfortably.
2. Use DifferentialIKController in **absolute pose mode**:
    - Command a static target pose in the center of your “good” workspace (e.g., x=0.22, y=−0.2, z=0.0).
3. Observe:
    - If it converges smoothly and holds position, your config is basically correct.
    - If it oscillates or fails, focus on:
        - Jacobian frame/order
        - `ik_method="dls"` and `lambda_val`
        - `command_type` / `use_relative_mode` consistency.

If this static test works but the interpolated trajectory fails, then the bug is almost certainly in:

- how you build the start/end interpolation vs. relative/absolute interpretation,
- or how you distribute trajectory updates across sim vs policy frequencies.



[^1]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.controllers.html

[^2]: https://isaac-sim.github.io/IsaacLab/main/source/tutorials/05_controllers/run_diff_ik.html

[^3]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.controllers.html

[^4]: https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.controllers.html#isaaclab.controllers.DifferentialIKControllerCfg

[^5]: https://arxiv.org/html/2511.04831v1

[^6]: https://github.com/isaac-sim/IsaacLab/issues/4971

[^7]: https://www.youtube.com/watch?v=WqNkmz0BnzQ

[^8]: https://forums.developer.nvidia.com/t/holonomic-controller-vs-differential-controller/300552

[^9]: https://github.com/isaac-sim/IsaacLab/discussions/1838

[^10]: https://www.youtube.com/watch?v=CUag-iyR2jo

[^11]: https://www.youtube.com/watch?v=Nwa7xP3RtO0

[^12]: https://www.reddit.com/r/reinforcementlearning/comments/1j4wa9g/which_robotics_simulator_is_better_for/

[^13]: https://forums.developer.nvidia.com/t/orbit-differentialikcontroller-for-velocity-control/280504

**Reply by InnaSviri** [reply]
1. I use trajectory in absolute Cartesian: command_type="position" (without orientation), use_relative_mode=False, command = x_interp(t) (no deltas). Desired EE position is in the same frame as ee_pos passed into compute.

2. I've done _compute_frame_jacobian and _compute_frame_pose similarly to DifferentialInverseKinematicsAction so they are in the same frame and convention.
sanity tect is fine - [video](https://1drv.ms/v/c/df93b9afd6f38bf0/IQDHMS5T_uU3Q41AumcxGArVAb9b_44LphGNC-vZIOyFsXg?e=a6eZGH)

3. Yes, I've experimented with a number of methods and parameters, and I've found that 'dls' with lambda=0.05 is the best one.  Here's my cfg:
<img width="814" height="60" alt="ik_cfg" src="https://github.com/user-attachments/assets/4ad0a28e-cae9-4c2a-96ee-c9a4dc86ba01" />

4. Am I correct in understanding that it is better to call ik.set_command at the simulation frequency, as well as ik.compute? I tried to set command more frequently - at sim frequency, but the result leaves a lot to be desired [video](https://1drv.ms/v/c/df93b9afd6f38bf0/IQDJ9nKYOP7zRqkqvm7TWU8gARvpcyLQB56Lhlrk8vdxQW0?e=yyh3um)

5. I clamp ik result to soft_joints_limits before sending
<img width="811" height="178" alt="joint_clamp" src="https://github.com/user-attachments/assets/866d0a85-ce5a-497b-bb59-4153320eb5e6" />
Yes! As I said before, the situation improves slightly with more limited sampling ranges when they are away from singular zones. Unfortunately, this is not enough to train a policy on such a dataset.
I tried increasing lambda_val to 0.07 and reducing the delta between x_interp(t1) and x_interp(t2), but the end-effector still flies off the target. :(

6.  Static test is fine - [video](https://1drv.ms/v/c/df93b9afd6f38bf0/IQBGPnJyn5wWRZQELQ0C03DtAcFOcabpijksqmYWpF3__zo?e=SgBPXB)

My trajectory sampling is rather straightforward:

`def _resample_command(self, env_ids):

        if len(env_ids) == 0:
            return
        self._adaptive_sampling(env_ids)

        if self.cfg.play:
            self.time_steps[env_ids] = 0

        self.t[env_ids] = 0

        device = self.device
        N = len(env_ids)
        T = self.cfg.time_step_total

        # Sample start pose
        pos_start_ranges = torch.tensor(
            [self.cfg.pose_range[k] for k in ["x_start", "y_start", "z_start"]],
            device=device,
        )
        pos_end_ranges = torch.tensor(
            [self.cfg.pose_range[k] for k in ["x_end", "y_end", "z_end"]],
            device=device,
        )
        ori_start_ranges = torch.tensor(
            [self.cfg.pose_range[k] for k in ["roll_start", "pitch_start", "yaw_start"]],
            device=device,
        )
        ori_end_ranges = torch.tensor(
            [self.cfg.pose_range[k] for k in ["roll_end", "pitch_end", "yaw_end"]],
            device=device,
        )

        start_pos = sample_uniform(
            pos_start_ranges[:, 0], pos_start_ranges[:, 1],
            (N, 3),
            device=device,
        )

        end_pos = sample_uniform(
            pos_end_ranges[:, 0], pos_end_ranges[:, 1],
            (N, 3),
            device=device,
        )

        start_euler = sample_uniform(
            ori_start_ranges[:, 0], ori_start_ranges[:, 1],
            (N, 3),
            device=device,
        )

        end_euler = sample_uniform(
            ori_end_ranges[:, 0], ori_end_ranges[:, 1],
            (N, 3),
            device=device,
        )

        # Convert Euler → quaternion
        start_quat = quat_from_euler_xyz(start_euler[..., 0], start_euler[..., 1], start_euler[..., 2])  # (N,4)
        end_quat = quat_from_euler_xyz(end_euler[..., 0], end_euler[..., 1], end_euler[..., 2])

        # Interpolation parameter
        t = torch.linspace(0.0, 1.0, T, device=device)

        # Position interpolation
        interp_pos = (
                start_pos[:, None, :] * (1 - t.view(1, T, 1))
                + end_pos[:, None, :] * t.view(1, T, 1)
        )

        # Orientation interpolation
        interp_quat = quaternion_slerp(start_quat, end_quat, t)

        self.ee_pos_traj[env_ids] = interp_pos
        self.ee_quat_traj[env_ids] = interp_quat

        self.ik.reset()`

Then, at either the SIM or the policy frequency (I've tried both), I do:

`self.ee_pos_command_b = self.ee_pos_traj[env_ids, self.time_steps]
      # obtain quantities from simulation
      ee_pos_curr, ee_quat_curr = self._compute_frame_pose()
       # set command into controller
      self.ik.set_command(torch.cat([self.ee_pos_command_b], dim=-1),
                                              ee_pos_curr, ee_quat_curr)`

I have tried various interpolation densities but that hasn't solved my problem.

Could you perhaps offer some insight into what might be causing the issue in my case?
